timelock-impl/src/main/java/com/palantir/atlasdb/timelock/management/AllNodesDisabledNamespacesUpdater.java

```diff
@@ -0,0 +1,271 @@
+        this.remoteUpdaters = remoteUpdaters;
+        this.remoteExecutors = remoteExecutors;
+        this.localUpdater = localUpdater;
+        this.lockIdSupplier = lockIdSupplier;
+        this.clusterSize = remoteUpdaters.size() + 1;
+    }
+
+    public static AllNodesDisabledNamespacesUpdater create(
+            AuthHeader authHeader,
+            ImmutableList<DisabledNamespacesUpdaterService> updaters,
+            Map<DisabledNamespacesUpdaterService, CheckedRejectionExecutorService> executors,
+            TimelockNamespaces localUpdater) {
+        return new AllNodesDisabledNamespacesUpdater(authHeader, updaters, executors, localUpdater, UUID::randomUUID);
+    }
+
+    public DisableNamespacesResponse disableOnAllNodes(Set<Namespace> namespaces) {
+        if (anyNodeIsUnreachable()) {
+            return DisableNamespaceResponses.unsuccessfulDueToPingFailure(namespaces);
+        }
+
+        UUID lockId = lockIdSupplier.get();
+        List<SingleNodeUpdateResponse> responses = disableNamespacesOnAllNodes(namespaces, lockId);
+        if (updateWasSuccessfulOnAllNodes(responses)) {
+            return DisableNamespacesResponse.successful(SuccessfulDisableNamespacesResponse.of(lockId));
+        }
+
+        Set<Namespace> consistentlyLockedNamespaces = getConsistentFailures(responses);
+        if (!consistentlyLockedNamespaces.isEmpty()) {
+            return DisableNamespaceResponses.unsuccessfulDueToConsistentlyLockedNamespaces(
+                    consistentlyLockedNamespaces);
+        }
+
+        // If no namespaces were consistently disabled, we should roll back our request,
+        // to avoid leaving ourselves in an inconsistent state.
+        boolean rollbackSuccess = attemptReEnableOnAllNodes(namespaces, lockId, responses.size());
+        if (rollbackSuccess) {
+            return DisableNamespaceResponses.unsuccessfulButRolledBack(namespaces, lockId);
+        }
+
+        return DisableNamespaceResponses.unsuccessfulAndRollBackFailed(namespaces, lockId);
+    }
+
+    public ReenableNamespacesResponse reEnableOnAllNodes(ReenableNamespacesRequest request) {
+        Set<Namespace> namespaces = request.getNamespaces();
+
+        if (anyNodeIsUnreachable()) {
+            return ReEnableNamespaceResponses.unsuccessfulDueToPingFailure(namespaces);
+        }
+
+        List<SingleNodeUpdateResponse> responses = reEnableNamespacesOnAllNodes(request);
+        if (updateWasSuccessfulOnAllNodes(responses)) {
+            return ReenableNamespacesResponse.successful(SuccessfulReenableNamespacesResponse.of(true));
+        }
+
+        Set<Namespace> consistentlyLockedNamespaces = getConsistentFailures(responses);
+        if (!consistentlyLockedNamespaces.isEmpty()) {
+            return ReEnableNamespaceResponses.unsuccessfulDueToConsistentlyLockedNamespaces(
+                    consistentlyLockedNamespaces);
+        }
+
+        // TODO(gs): should only attempt to roll back locally?
```

Comment:
I don't think we ever want to re-disable if re-enable fails. We should just unlock as much as we can and report any potential failures. We should only do re-enabling when it's safe to do so, so there is no benefit to re-disabling and it could potentially introduce bad race conditions that we could have avoided

Reply:
sure, fine by me - that's why I kept the two workflows separate.